datalake: introduce partition translator

bharathv opened the pull request:
A partition translator is responsible for translating data from a given
partition periodically. A single instance of translator should be active for
a given partition and is attached to the leader replica.

The translator periodically wakes up, checks if there is new data that is
pending translation and creates a reader for the data range and hands it off
to low level iceberg translator that converts the data.

The frequency is controlled via topic property `redpanda.iceberg.translation.interval.ms`.

Added a cluster fixture that wires up an e2e test with basic validation.

More tests are coming in future patches once we have a ducktape service running.


<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [x] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.2.x
- [ ] v24.1.x
- [ ] v23.3.x

## Release Notes
* none

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

bharathv commented:
/dt

bharathv commented:
/dt

bharathv commented:
/dt

vbotbuildovich commented:
<!--NEW FAILURES-->
new failures in https://buildkite.com/redpanda/redpanda/builds/55859#01925a91-0a93-45be-8ea1-96c62422c2f7:
```
"rptest.tests.archive_retention_test.CloudArchiveRetentionTest.test_delete.cloud_storage_type=CloudStorageType.ABS.retention_type=retention.ms"
"rptest.tests.cluster_recovery_test.ClusterRecoveryTest.test_bootstrap_with_recovery"
"rptest.tests.e2e_topic_recovery_test.EndToEndTopicRecovery.test_restore_with_config_batches.num_messages=2.cloud_storage_type=CloudStorageType.S3"
"rptest.tests.e2e_shadow_indexing_test.EndToEndShadowIndexingTest.test_reset.cloud_storage_type=CloudStorageType.ABS"
"rptest.tests.e2e_shadow_indexing_test.EndToEndShadowIndexingTest.test_reset_spillover.cloud_storage_type=CloudStorageType.ABS"
"rptest.tests.e2e_shadow_indexing_test.EndToEndThrottlingTest.test_throttling.cloud_storage_type=CloudStorageType.S3"
"rptest.tests.shadow_indexing_compacted_topic_test.ShadowIndexingCompactedTopicTest.test_upload.cloud_storage_type=CloudStorageType.S3"
"rptest.tests.topic_delete_test.TopicDeleteCloudStorageTest.drop_lifecycle_marker_test.cloud_storage_type=CloudStorageType.ABS"
"rptest.tests.topic_recovery_test.TopicRecoveryTest.test_many_partitions.cloud_storage_type=CloudStorageType.ABS.check_mode=check_manifest_and_segment_metadata"
"rptest.tests.topic_recovery_test.TopicRecoveryTest.test_many_partitions.cloud_storage_type=CloudStorageType.S3.check_mode=check_manifest_existence"
"rptest.tests.archival_test.ArchivalTest.test_timeboxed_uploads.acks=0.cloud_storage_type=CloudStorageType.ABS"
"rptest.tests.delete_records_test.DeleteRecordsTest.test_delete_records_segment_deletion.cloud_storage_enabled=True.truncate_point=at_high_watermark"
"rptest.tests.delete_records_test.DeleteRecordsTest.test_delete_records_segment_deletion.cloud_storage_enabled=True.truncate_point=start_offset"
"rptest.tests.retention_policy_test.ShadowIndexingCloudRetentionTest.test_cloud_time_based_retention.cloud_storage_type=CloudStorageType.ABS"
"rptest.tests.retention_policy_test.ShadowIndexingCloudRetentionTest.test_cloud_size_based_retention.cloud_storage_type=CloudStorageType.ABS"
```
<!--74c38e38-9c01-44de-a201-f2a350570d5e-->

vbotbuildovich commented:
<!--RETRIED FAILURES-->
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/55859#01925a91-0a93-45be-8ea1-96c62422c2f7
<!--1fd7ad85-889a-412c-9044-10f1aacd3d09-->

bharathv commented:
/dt

bharathv commented:
/dt

bharathv commented:
/dt

vbotbuildovich commented:
<!--UNIT FAILURES-->
<!--0ba1f8a6-f8d7-4f29-97af-ac831b6559f7-->
<!--body-->
the below tests from https://buildkite.com/redpanda/redpanda/builds/56471#01928bbb-7d35-452e-b1f6-590ee059a83a have failed and will be retried
```
kafka_server_rpfixture
```
<!--/body-->

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/56471-->
<!--857c2da9-a5c6-4ccb-8ccc-db73ef16dea1-->
<!--title-->
## Retry command for [Build#56471](https://buildkite.com/redpanda/redpanda/builds/56471)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/cloud_storage_timing_stress_test.py::CloudStorageTimingStressTest.test_cloud_storage@{"cleanup_policy":"delete"}
```
<!--/body-->

bharathv commented:
/dt

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/56494-->
<!--f107744b-763d-4f08-a500-617b02e2a336-->
<!--title-->
## Retry command for [Build#56494](https://buildkite.com/redpanda/redpanda/builds/56494)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/alter_topic_configuration_test.py::AlterConfigMixedNodeTest.test_alter_config_shadow_indexing_mixed_node@{"incremental_update":true}
```
<!--/body-->

bharathv commented:
Failures are unrelated to the PR

* bazel is broken because the writer/avro is not fully bazel friendly.
* test_alter_config_shadow_indexing_mixed_node - I think thats a test [issue](https://redpandadata.slack.com/archives/C07FJGU5AKV/p1728974396211999).
* test_snapshot_compatibility - is unrelated, I see it happening on other PRs too, I'll take a look at it separately.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/56516-->
<!--bccd7ad2-7e0f-4abf-8112-b2e875bcf195-->
<!--title-->
## Retry command for [Build#56516](https://buildkite.com/redpanda/redpanda/builds/56516)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/alter_topic_configuration_test.py::AlterConfigMixedNodeTest.test_alter_config_shadow_indexing_mixed_node@{"incremental_update":false}
```
<!--/body-->

mmaslankaprv commented on src/v/storage/segment_utils.h:
```diff
@@ -238,6 +238,13 @@ is_compactible_control_batch(const model::record_batch_type batch_type) {
 }
 
+inline bool
+compactible_offset_translator_batch(const model::record_batch_type batch_type) {
+    // certain batches not contributing to the kafka offset space are
+    // compactible
+    return batch_type == model::record_batch_type::datalake_translation_state;
+}
+
```
i think this will lead to an issue in which we may result in an offset inconsistency. Offset translator state is not checkpointed to kvstore on every update, if those batches are compacted they won't contribute to the offset translator state anymore leading to Kafka offsets shift.

bharathv replied:
good catch, I think it would also mess up partition moves after compaction as the new node builds OTS based on the compacted log. will remove this commit.

rockwotj replied:
naive question: is there a way to catch this with testing in the future? This ought to be a scenario we can unit test in theory.

bharathv replied:
https://github.com/redpanda-data/redpanda/pull/23817

mmaslankaprv commented on src/v/datalake/translation/state_machine.cc:
```diff
@@ -0,0 +1,147 @@
+      .highest_translated_offset = offset};
+    storage::record_batch_builder builder(
+      model::record_batch_type::datalake_translation_state, model::offset(0));
+    builder.add_raw_kv(std::nullopt, serde::to_iobuf(val));
```
i think setting a key to null may impact compacted topics, but i think we should not compact batches that contribute to offset translation.

mmaslankaprv commented on src/v/datalake/translation/state_machine.cc:
```diff
@@ -0,0 +1,147 @@
+      != model::record_batch_type::datalake_translation_state) {
+        co_return;
+    }
+    auto records = batch.copy_records();
```
nit: there is a nice record iterator interface that allows doing this apply loop generically, i know that there will always be a single record but it allow skipping vassert here

mmaslankaprv commented on src/v/datalake/translation/state_machine.cc:
```diff
@@ -0,0 +1,147 @@
+    if (!co_await sync(timeout) || _insync_term < term) {
+        co_return raft::errc::not_leader;
+    }
+    vlog(
+      datalake_log.debug,
+      "sync_with_cooordinator {}, {}",
+      new_translated_offset,
+      term);
```
this log line will be very hard to read, can you add what is being logged

mmaslankaprv commented on src/v/datalake/translation/state_machine.cc:
```diff
@@ -0,0 +1,147 @@
+  model::term_id term,
+  model::timeout_clock::duration timeout,
+  ss::abort_source& as) {
+    if (!co_await sync(timeout) || _insync_term < term) {
```
should the check be `_insync_term != term`

mmaslankaprv commented on src/v/datalake/translation/state_machine.cc:
```diff
@@ -0,0 +1,147 @@
+    auto& r = *records.begin();
+    auto val = serde::from_iobuf<translation_state>(r.release_value());
+    vlog(
+      datalake_log.trace,
```
persisted stm has logger `_log` which adds additional context to log lines.

mmaslankaprv commented on src/v/datalake/translation/partition_translator.cc:
```diff
@@ -0,0 +1,273 @@
+kafka::offset partition_translator::max_offset_for_translation() const {
+    // We factor in LSO to ensure only committed transactional batches are
+    // translated and we clamp it to committed offset to avoid translation
+    // of majorty appended but not flushed data (eg: acks=0/1/write_caching).
```
nit: typo - `majority`

mmaslankaprv commented on src/v/datalake/translation/partition_translator.cc:
```diff
@@ -0,0 +1,273 @@
+          _partition->last_stable_offset());
+        co_return translation_success::yes;
+    }
+    // We have some data to translate, make a reader
+    // and dispatch to the iceberg translator
+    //
+    // todo: what if we have too little to read, eg: slow moving topic.
+    // Currently we rely on user setting the translation interval
+    // appropriately to ensure each attempt has something substantial
+    // to read. We can add some heuristics to estimate the size and
+    // backoff if it is too small and wait for enough data to
+    // accumulate. The resulting parquet files are only performant
+    // if there is a big chunk of data in them. Smaller parquet files
```
I think here we can use `disk_log_impl::size_bytes_after_offset` to first query the possible read size, this way we can know for sure what the read size is. I think this is required for v1 as without that we may end up with lots of tiny parque files.

bharathv replied:
right, I did consider `disk_io_impl::offset_range_size()` which seems like a more accurate version of size_bytes_after_offset(), left it out for v1, we still have to give up at some point right?  

eg: there is only little data to read but it has been long time since since last translation at which point we should probably just translate even if the data is small.

bharathv replied:
chatted offline: this heuristic will be in a separate patch.

mmaslankaprv commented on src/v/datalake/translation/partition_translator.h:
```diff
@@ -0,0 +1,166 @@
+    // how many parallel translations can run at one point as we operate under
+    // a memory budget for all translations (semaphore below).
+    size_t _max_bytes_per_reader;
+    ssx::semaphore& _parallel_translations;
```
consider using a pointer for consistency ?

mmaslankaprv replied:
also if this is global per shard maybe we can wrap sempahore in a sharded object, this way it will be easier to add observability in future.

mmaslankaprv commented on src/v/datalake/translation/partition_translator.h:
```diff
@@ -0,0 +1,166 @@
+            while (num_attempts-- > 0) {
+                auto result = co_await ss::futurize_invoke(f);
+                if (num_attempts > 0 && should_retry(result)) {
+                    co_await ss::sleep(next());
```
abortable ?

mmaslankaprv commented on src/v/datalake/translation/partition_translator.cc:
```diff
@@ -0,0 +1,273 @@
+      &proxy, std::move(log_reader.ot_state));
+    auto kafka_reader = kafka::read_committed_reader(
+      std::move(tracker), std::move(log_reader.reader));
+    auto units = co_await ss::get_units(_parallel_translations, 1, _as);
```
maybe we can grab units a bit earlier, when we actually create a reader not to allocate any resources. Now the readear is read and waiting for the semaphore

bharathv replied:
I can move it a few lines above but the reader is lazy right? It doesn't read anything until L175+ where we pass a consumer?

mmaslankaprv commented on src/v/datalake/translation/partition_translator.cc:
```diff
@@ -0,0 +1,273 @@
+      [this](coordinator::add_translated_data_files_reply result) {
+          return can_continue() && is_retriable(result.errc);
+      });
+    vlog(_logger.trace, "Adding tranlsated data file, result: {}", result);
```
nit :-)

mmaslankaprv commented on src/v/datalake/translation/partition_translator.cc:
```diff
@@ -0,0 +1,273 @@
+             ->stm_manager()
+             ->get<datalake::translation::translation_stm>();
+    vassert(_stm, "No translation stm found for {}", _partition->ntp());
+    ssx::repeat_until_gate_closed(_gate, [this, sg] {
+        return ss::with_scheduling_group(sg, [this] { return do_translate(); });
+    });
```
consider separating this into a `start()` method to be able to decouple creation from starting

mmaslankaprv commented on src/v/datalake/datalake_manager.cc:
```diff
@@ -38,10 +41,127 @@ datalake_manager::datalake_manager(
+  , _effective_max_translator_buffered_data(
+      std::min(memory_limit, max_translator_buffered_data))
+  , _parallel_translations(
+      size_t(
+        std::floor(memory_limit / _effective_max_translator_buffered_data)),
+      "datalake_parallel_translations")
```
nit: consider moving this logic into a separate method

mmaslankaprv commented on src/v/datalake/datalake_manager.cc:
```diff
@@ -38,10 +41,127 @@ datalake_manager::datalake_manager(
+    co_await std::move(f);
+}
+
+void datalake_manager::on_group_notifiction(raft::group_id group) {
```
nit: typo

mmaslankaprv commented on src/v/datalake/datalake_manager.cc:
```diff
@@ -38,10 +41,127 @@ datalake_manager::datalake_manager(
+        ssx::spawn_with_gate(
+          _gate, [this, partition] { return start_translator(partition); });
+    } else {
+        auto default_value
+          = config::shard_local_cfg().iceberg_translation_interval_ms_default();
```
binding ?

mmaslankaprv replied:
or maybe use the metadata cache the same way as other topic properties

mmaslankaprv commented on src/v/datalake/datalake_manager.h:
```diff
@@ -47,18 +53,37 @@ class datalake_manager : public ss::peering_sharded_service<datalake_manager> {
 
 private:
+    using translator = std::unique_ptr<translation::partition_translator>;
+    using translator_map = absl::btree_map<raft::group_id, translator>;
```
does it have to be a `btree` or can we use `chunked_hashmap` ?

bharathv replied:
it doesn't have to be, will switch.

bharathv commented on src/v/datalake/translation/partition_translator.cc:
```diff
@@ -0,0 +1,273 @@
+    // accumulate. The resulting parquet files are only performant
+    // if there is a big chunk of data in them. Smaller parquet files
+    // are an overhead for iceberg metadata.
+    auto proxy = kafka::make_partition_proxy(_partition);
```
transform/api also does make_partition_proxy(), I think the reason is kafka::aborted_transaction_tracker() is wired up with a partition_proxy, which seems correct so one could read cloud storage data too via txn_reader.

Probably in the future if we have to hydrate data from TS to datalake, this may come handy but at this point it seems a bit heavy as you said.

bharathv commented on src/v/datalake/datalake_manager.cc:
```diff
@@ -38,10 +41,127 @@ datalake_manager::datalake_manager(
+  , _translation_ms_conf(config::shard_local_cfg()
+                           .iceberg_translation_interval_ms_default.bind()) {
+    // Handle leadership changes
+    auto leadership_registration
+      = _group_mgr->local().register_leadership_notification(
+        [this](
+          raft::group_id group,
+          ::model::term_id,
+          std::optional<::model::node_id>) { on_group_notifiction(group); });
```
ya partition moving cross cores without leadership change seems like a problem, will wire up managed notifications too.

> To make sure we clean up all the translators when partition is stopped on the core.

I think that will happen in the current setup too as the local consensus instance is no longer the leader

rockwotj replied:
I think I had to registry for both `partition_manager` and `group_manager` in transforms. I'd assume that's true here?

rockwotj commented on src/v/kafka/server/handlers/topics/types.h:
```diff
@@ -102,6 +102,9 @@ inline constexpr std::string_view topic_property_leaders_preference
 inline constexpr std::string_view topic_property_cloud_topic_enabled
   = "redpanda.cloud_topic.enabled";
 
+static constexpr std::string_view topic_property_iceberg_translation_interval_ms
```
inline

rockwotj commented on src/v/datalake/translation/BUILD:
```diff
@@ -0,0 +1,14 @@
+        "//src/v/model",
+        "//src/v/serde",
+    ],
+)
```
nit: newline

rockwotj commented on src/v/datalake/translation/state_machine.h:
```diff
@@ -0,0 +1,64 @@
+
+/// Tracks the progress of datalake translation and clamps the collectible
+/// offset to ensure the data is not GC-ed before translation is finished.
+class translation_stm final : public raft::persisted_stm<> {
```
How much effort would it be to setup a unit test for this class?

rockwotj commented on src/v/datalake/translation/partition_translator.h:
```diff
@@ -0,0 +1,166 @@
+
+/**
+ * A partition translator is responsible for translating data from a given
+ * partition periodically. A single instance of translator should be active for
+ * a given partition and is attached to the leader replica.
```
Should we callout that it's possible during leadership transitions for there to be two translators?

rockwotj commented on src/v/datalake/translation/partition_translator.h:
```diff
@@ -0,0 +1,166 @@
+ *
+ * It is possible that the translation stm can get out of sync with what is
+ * on the coordinator, so it is reconciled as needed. It is ok to work on a
+ * stale stm state as the eventual checkpoint request with the coordinator
+ * fenced based on the translator term.
```
this reads oddly

rockwotj commented on src/v/datalake/translation/partition_translator.h:
```diff
@@ -0,0 +1,166 @@
+    void reset_translation_interval(std::chrono::milliseconds new_base);
+
+private:
+    struct backoff {
```
Does this struct need to go in the header? Can we hide in `cc` file?

rockwotj commented on src/v/datalake/translation/partition_translator.h:
```diff
@@ -0,0 +1,166 @@
+                }
+                co_return result;
+            }
+            __builtin_unreachable();
```
Is this really unreachable? It seems like if you are retrying something it's valid to exhaust retries - should we really abort the process? At least `vassert` for better logging...

rockwotj replied:
I guess the intention is that `should_retry` should return false before `max_attempts` is added. Should we rewrite the loop so that the behavior is the same if someone is a little careless or we have a logic bug in `should_retry`? May just throw an exception.

rockwotj commented on src/v/datalake/translation/partition_translator.cc:
```diff
@@ -0,0 +1,273 @@
+}
+
+kafka::offset partition_translator::min_offset_for_translation() const {
+    auto min_offset = _partition->raft_start_offset();
+    return model::offset_cast(
+      _partition->get_offset_translator_state()->from_log_offset(min_offset));
```
There are a couple of palaces we have to remember to translate state - the different type helps a little here, but should we just keep a partition_proxy instead of a raw cluster::partition member in this class? IIRC that will handle all the translation, which it seems good to try and keep that in a single layer.

rockwotj commented on src/v/datalake/translation/partition_translator.h:
```diff
@@ -0,0 +1,166 @@
+ *    sync_stm_with_coordinator(md)
+ */
+
+class partition_translator {
```
How much effort is it to get a unit test for this? At least laydown the foundations for better tests later (I guess unlikely, but is nice to hope!).

rockwotj commented on src/v/datalake/datalake_manager.cc:
```diff
@@ -42,8 +44,124 @@ datalake_manager::datalake_manager(
+        return;
+    }
+    auto it = _translators.find(group);
+    // todo handle topic / partition disabling
```
nit: I generally prefer doing something like `TODO(iceberg)` so it's clearly grouped to the project and makes it easy to grep for all the iceberg todos before we GA 😄

rockwotj commented on src/v/datalake/translation/CMakeLists.txt:
```diff
@@ -0,0 +1,11 @@
+        v::model
+        v::raft
+        Seastar::seastar
+)
```
nit: newline

rockwotj reviewed (commented):
Still need to look a little more at the tests (last commit), but wanted to leave current feedback. Nice work here!

andrwng commented on src/v/datalake/translation/state_machine.cc:
```diff
@@ -0,0 +1,147 @@
+      datalake_log.trace,
+      "updating highest translated offset to {}",
+      val.highest_translated_offset);
+    _highest_translated_offset = val.highest_translated_offset;
```
Should we std::max these? If other code remains bug free forever I imagine it makes no difference, but that's a big if

andrwng commented on src/v/datalake/translation/state_machine.cc:
```diff
@@ -0,0 +1,147 @@
+    co_return _highest_translated_offset;
+}
+
+ss::future<std::error_code> translation_stm::sync_with_coordinator(
```
nit: from the name it's easy to incorrectly assume that this calls out to the coordinator. Maybe call this `reset_highest_translated_offset()` or something?

andrwng commented on src/v/datalake/translation/partition_translator.cc:
```diff
@@ -0,0 +1,273 @@
+  , _writer_scratch_space(std::filesystem::temp_directory_path())
+  , _logger(prefix_logger{
+      datalake_log, fmt::format("{}-term-{}", _partition->ntp(), _term)}) {
+    _stm = _partition->raft()
+             ->stm_manager()
+             ->get<datalake::translation::translation_stm>();
```
nit: just curious why this isn't in the initializer list

andrwng commented on src/v/datalake/translation/partition_translator.cc:
```diff
@@ -0,0 +1,273 @@
+          [this] { return do_translate_once(); },
+          [this](translation_success result) {
+              auto retry = can_continue() && result == translation_success::no;
+              if (retry) {
+                  _reconcile = needs_reconciliation::yes;
+              }
```
I'm wondering if it makes sense to just always try to sync with the coordinator on failure within do_translate_once -- then we wouldn't need to keep track of _reconcile, and the happy path of do_translate could always rely on the STM

andrwng commented on src/v/datalake/translation/partition_translator.cc:
```diff
@@ -0,0 +1,273 @@
+    // We factor in LSO to ensure only committed transactional batches are
+    // translated and we clamp it to committed offset to avoid translation
+    // of majorty appended but not flushed data (eg: acks=0/1/write_caching).
+    auto max_log_offset = std::min(
+      model::prev_offset(_partition->last_stable_offset()),
+      _partition->committed_offset());
+    return model::offset_cast(
+      _partition->get_offset_translator_state()->from_log_offset(
+        max_log_offset));
```
nit: wondering if this is code we can get from the Kafka partition?

asimms41 commented on src/v/config/configuration.cc:
```diff
@@ -3707,6 +3707,15 @@ configuration::configuration()
+  , iceberg_translation_interval_ms_default(
+      *this,
+      "iceberg_translation_interval_ms_default",
+      "Controls how often an iceberg enabled topic is checked for new data to "
```
```suggestion
      "Controls how often an Iceberg enabled topic is checked for new data to "
```

asimms41 reviewed (commented):
Just a minor change

Deflaimun commented on src/v/config/configuration.cc:
```diff
@@ -3707,6 +3707,15 @@ configuration::configuration()
+      *this,
+      "iceberg_translation_interval_ms_default",
+      "Controls how often an iceberg enabled topic is checked for new data to "
+      "translate. redpanda.iceberg.translation.interval.ms is the topic level "
```
@bharathv which takes precedence? The cluster config or the topic config?

dotnwat reviewed (commented):
this commit doesn't compile

> datalake/tests: add a cluster test fixture with datalake enabled.

the error

```
nwatkins@acid:~/src/redpanda-bisect-check$ cat 23633.stderr.txt
INFO: Invocation ID: e2f6926d-c2e6-48cd-99fa-994f3387f135
DEBUG: /home/nwatkins/.cache/bazel/_bazel_nwatkins/27f5827f173f731481084ebbdf273b17/external/rules_pkg~/pkg/mappings.bzl:310:34: same source mapped to different locations <generated file src/v/redpanda/redpanda> redpanda redpanda.runfiles/_main/src/v/redpanda/redpanda
INFO: Analyzed 1168 targets (18 packages loaded, 1124 targets configured).
ERROR: /home/nwatkins/src/redpanda-bisect-check/redpanda/src/v/datalake/translation/BUILD:35:20: Compiling src/v/datalake/translation/partition_translator.cc failed: (Exit 1): cc_wrapper.sh failed: error executing CppCompile command (from target //src/v/datalake/translation:partition_translator) external/toolchains_llvm~~llvm~llvm_18_toolchain/bin/cc_wrapper.sh -U_FORTIFY_SOURCE '--target=x86_64-unknown-linux-gnu' -U_FORTIFY_SOURCE -fstack-protector -fno-omit-frame-pointer -fcolor-diagnostics ... (remaining 925 arguments skipped)

Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
src/v/datalake/translation/partition_translator.cc:15:10: fatal error: 'datalake/batching_parquet_writer.h' file not found
   15 | #include "datalake/batching_parquet_writer.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```

bharathv commented:
> this commit doesn't compile

@dotnwat this is the part that requires arrow-bazel integration.

dotnwat commented:
> > this commit doesn't compile
> 
> @dotnwat this is the part that requires arrow-bazel integration.

got it. is this pr blocked then on arrow-bazel, or do we want to remove something from the bazel build in this pr?

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/56955-->
<!--07ecd5b1-2525-4132-8a93-2e672e715f59-->
<!--title-->
## Retry command for [Build#56955](https://buildkite.com/redpanda/redpanda/builds/56955)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/alter_topic_configuration_test.py::AlterConfigMixedNodeTest.test_alter_config_shadow_indexing_mixed_node@{"incremental_update":true}
```
<!--/body-->

bharathv commented:
Ducktape test failures are unrelated, I've split it out into a separate [PR](https://github.com/redpanda-data/redpanda/pull/23874) just to make sure I'm not breaking something else.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/56980-->
<!--f7d39906-c34d-42b1-b781-7f733c59eac9-->
<!--title-->
## Retry command for [Build#56980](https://buildkite.com/redpanda/redpanda/builds/56980)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/alter_topic_configuration_test.py::AlterConfigMixedNodeTest.test_alter_config_shadow_indexing_mixed_node@{"incremental_update":true}
```
<!--/body-->

bharathv commented:
/ci-repeat 1
dt-repeat 10
tests/rptest/tests/alter_topic_configuration_test.py::AlterConfigMixedNodeTest.test_alter_config_shadow_indexing_mixed_node

dotnwat commented:
This commit

> cmake: tidyup dependency graph

fails to build in bazel

```
src/v/datalake/translation/partition_translator.cc:15:10: fatal error: 'datalake/batching_parquet_writer.h' file not found
   15 | #include "datalake/batching_parquet_writer.h"
```

bharathv commented:
> This commit
> 
> > cmake: tidyup dependency graph
> 
> fails to build in bazel
> 
> ```
> src/v/datalake/translation/partition_translator.cc:15:10: fatal error: 'datalake/batching_parquet_writer.h' file not found
>    15 | #include "datalake/batching_parquet_writer.h"
> ```

hmm.. I fixed it in the commit right after, split for reviewer's convenience.. I'll fix-up the commit history then.

dotnwat commented:
>  fixed it in the commit right after, split for reviewer's convenience.. I'll fix-up the commit history then.

thanks, i appreciate the former. i guess `git bisect` appreciates the later :)

andrwng commented on src/v/datalake/datalake_manager.cc:
```diff
@@ -41,9 +46,159 @@ datalake_manager::datalake_manager(
+
+void datalake_manager::on_group_notification(const model::ntp& ntp) {
+    auto partition = _partition_mgr->local().get(ntp);
+    if (!partition || !model::is_user_topic(ntp)) {
```
Actually the common case is that it is a user topic, so probably not much of a win to swap

bharathv commented on src/v/redpanda/application.cc:
```diff
@@ -1995,6 +1996,9 @@ void application::wire_up_redpanda_services(
           std::ref(cloud_io),
           cloud_configs.local().bucket_name)
           .get();
+        _datalake_coordinator_mgr
+          .invoke_on_all(&datalake::coordinator::coordinator_manager::start)
+          .get();
```
cool, will remove.

bharathv commented on src/v/datalake/translation/state_machine.cc:
```diff
@@ -0,0 +1,147 @@
+        co_return raft::errc::success;
+    }
+    if (!as.abort_requested() && _raft->term() == current_term) {
+        co_await _raft->step_down("datalake coordinator sync error");
```
if this check fails (lets assume wait_no_throw() timed out for some hypothetical reason due to a delay in the applier, but raft is up to date), the state machine is in an inconsistent state from the log, meaning there are entries that are raft committed but not applied in the state machine.

So for a class of state machine implementations (which the code calls ingestion time stms), that look up the state machine more applying a change they operate on the inconsistent state. In this case a subsequent reset_highest_translated_offset() can potentially perform an incorrect check in L86, so this step down resets the state (as subsequent sync fails). Normally this is accompanied by a lock so this whole state change in  the lock scope but we have only one reader/writer for this stm.

andrwng replied:
I see, thanks for the context. I guess generally this helps avoid surprise updates to the STM when the decisions that build the updates are based on stale data. OTOH, it seems like this is also "handled" by making the STM robust to stale updates with the std::max() call (this is also somewhat what I'm doing in the coordinator, which applies a rough CAS)

No need to change anything, but just wondering if the std::max() in apply is enough now

bharathv replied:
right.. I think coordinator replicates first and reconciles later which the "ingestion time stm" tries to avoid as in some cases (eg: produce path) it is expensive to replicate, rather it resorts to tricks like this to ensure the checks always happen on the latest snapshot of data.

bharathv commented on src/v/datalake/translation/partition_translator.cc:
```diff
@@ -0,0 +1,348 @@
+partition_translator::do_translate_once(retry_chain_node& parent_rcn) {
+    if (
+      !_partition->get_ntp_config().iceberg_enabled()
+      || !_features->local().is_active(features::feature::datalake_iceberg)) {
```
I think the following scenario is possible.

- 3 node cluster, mixed mode, 1 node is upgraded to RP version with iceberg support and hosts the controller leader
- User creates an iceberg enabled topic and the upgrade node also has data partition leader.
- If we do not have this check in place, we end up replicating the "translation_state" batch to the followers on a lower version.
- If we only do this check in datalake manger and if the feature is not enabled by the time a translator instance is created and the leader never moves from the node, we risk missing a notification unless we periodically check for it (which is what this logic does).

andrwng replied:
But presumably we'd prevent iceberg topics from being created until the feature says that the whole cluster supports it, no? So step 2 would be blocked?

bharathv replied:
I don't think we do (not just for iceberg but for any feature), my guess is it is generally preferred to push the mixed mode complexity (wherever possible) into the feature itself rather than erroring out client requests until the upgrade finishes.

andrwng replied:
That's very interesting. Especially for net new features, I think it will easier to reason about the entire feature if we gate as early as possible, because reasoning about it downstream leaves around a ton of ambiguous edge cases (e.g. why gate here instead of in the datalake manager? are we going to create the coordinator in mixed mode? etc).

Anyway, not blocking -- we can always revisit this outside of this PR

bharathv replied:
right definitely easier to reason about if we don't let the topics into the system.

bharathv commented on src/v/datalake/translation/partition_translator.cc:
```diff
@@ -0,0 +1,348 @@
+        vlog(_logger.debug, "iceberg config/feature disabled, nothing to do.");
+        co_return translation_success::yes;
+    }
+    auto reconcile_result = co_await reconcile_with_coordinator();
+    if (!reconcile_result) {
+        vlog(_logger.warn, "reconciliation with coordinator failed");
+        co_return translation_success::no;
+    }
```
ya thats what the original version of the code did, I just removed it to simplify the code since this is an RPC that is done every 10s (or possibly higher) or so per partition (with iceberg enabled), seemed like a premature optimization but very easy to add back at a later time if needed.

bharathv commented on src/v/datalake/translation/partition_translator.cc:
```diff
@@ -0,0 +1,348 @@
+
+ss::future<> partition_translator::do_translate() {
+    while (can_continue()) {
+        co_await ss::sleep_abortable(_jitter.next_duration(), _as);
```
good idea.. I think we can do _as.check().

andrwng commented on src/v/redpanda/application.cc:
```diff
@@ -2024,10 +2023,14 @@ void application::wire_up_redpanda_services(
           sched_groups.datalake_sg(),
           memory_groups().datalake_max_memory())
           .get();
+        _datalake_manager.invoke_on_all(&datalake::datalake_manager::start)
+          .get();
```
Actually I ran into this problem because I didn't want to store sharded<> services in the coordinator manager. You should be fine here

bharathv commented on src/v/datalake/translation/partition_translator.cc:
```diff
@@ -0,0 +1,349 @@
+    parent_rcn.check_abort();
+    auto rcn = retry_chain_node(&parent_rcn);
+    while (true) {
+        auto result_f = co_await ss::coroutine::as_future<RetValueType>(
+          ss::futurize_invoke(f));
+        auto failed = result_f.failed();
+        // eagerly take the exception out to avoid ignored exceptional
+        // futures as retry() below can throw.
+        std::exception_ptr ex = failed ? result_f.get_exception() : nullptr;
+        auto retry = rcn.retry();
+        if (!retry.is_allowed) {
+            // No more retries allowed, propagated whatever we have.
+            if (failed) {
+                vassert(
+                  ex != nullptr,
+                  "Invalid exception, should be non null on a failed future.");
+                std::rethrow_exception(ex);
+            }
+            co_return result_f.get();
+        }
+        // Further retries are allowed, check for exceptions if any.
+        if (!failed) {
+            auto result = result_f.get();
+            if (!should_retry(result)) {
+                co_return result;
+            }
+        }
+        co_await ss::sleep_abortable(retry.delay, *retry.abort_source);
+    }
```
I wrote it the current way to have fewer conditionals, maybe I overthought this :)

The documentation of retry() specifically said this, hence a retry in a happy path, anyway I can see the flow is a bit not natural, will do a followup.

```
    /// The retry can be allowed or disallowed. The caller can call this
    /// method before the actual retry is needed or after.
```

andrwng replied:
Yea not a huge deal at all though FWIW seems like the same number of if blocks :D

Right, the retry_chain_node interface is flexible in that way, but it still reads a bit funny to someone unfamiliar with how it works. Maybe it's just a matter of picking a better name than `retry()`.

mmaslankaprv commented on src/v/datalake/coordinator/types.h:
```diff
@@ -58,12 +65,27 @@ struct add_translated_data_files_request
+        add_translated_data_files_request result;
+        result.tp = tp;
+        for (auto& range : ranges) {
+            result.ranges.push_back(range.copy());
```
nit: reserve ranges

mmaslankaprv commented on src/v/datalake/translation/partition_translator.h:
```diff
@@ -0,0 +1,130 @@
+    ss::shared_ptr<translation_stm> _stm;
+    ss::sharded<coordinator::frontend>* _frontend;
+    ss::sharded<features::feature_table>* _features;
+    std::unique_ptr<datalake::cloud_data_io>* _cloud_io;
```
nit: i think it will be more natural to use a plain pointer here, we do not need an ownership and lifecycle management in here

mmaslankaprv commented on src/v/datalake/coordinator/state_machine.cc:
```diff
@@ -136,8 +136,10 @@ coordinator_stm::apply_local_snapshot(raft::stm_snapshot_header, iobuf&&) {
+    // temporarily ignore snapshots, to be fixed later.
+    // throwing here results in uncaught exceptions, so a dummy snapshot
+    // at offset 0 avoids that.
+    co_return raft::stm_snapshot::create(0, model::offset{0}, iobuf{});
```
iiuc, now when iceberg is enabled it prevents local retention, we need to remember to fix this

mmaslankaprv reviewed (approved):
lgtm, nits only we can address them in subsequent PRs
